File: v1_16/src/main/java/me/fullidle/ficore/ficore/v1_16/BattleManager.java
package me.fullidle.ficore.ficore.v1_16;

import com.google.common.collect.Lists;
import com.pixelmonmod.pixelmon.Pixelmon;
import com.pixelmonmod.pixelmon.api.battles.BattleType;
import com.pixelmonmod.pixelmon.api.events.battles.BattleStartedEvent;
import com.pixelmonmod.pixelmon.api.pokemon.Pokemon;
import com.pixelmonmod.pixelmon.api.storage.StorageProxy;
import com.pixelmonmod.pixelmon.battles.BattleQuery;
import com.pixelmonmod.pixelmon.battles.BattleRegistry;
import com.pixelmonmod.pixelmon.battles.api.rules.BattleRules;
import com.pixelmonmod.pixelmon.battles.controller.BattleController;
import com.pixelmonmod.pixelmon.battles.controller.participants.*;
import com.pixelmonmod.pixelmon.entities.npcs.NPCEntity;
import com.pixelmonmod.pixelmon.entities.npcs.NPCTrainer;
import com.pixelmonmod.pixelmon.entities.pixelmon.PixelmonEntity;
import lombok.Getter;
import lombok.SneakyThrows;
import lombok.val;
import me.fullidle.ficore.ficore.common.V1_version;
import me.fullidle.ficore.ficore.common.api.data.FIData;
import me.fullidle.ficore.ficore.common.api.pokemon.battle.IBattleManager;
import me.fullidle.ficore.ficore.common.api.pokemon.battle.IPokeBattle;
import me.fullidle.ficore.ficore.common.api.pokemon.battle.actor.Actor;
import me.fullidle.ficore.ficore.common.api.pokemon.battle.actor.ActorManager;
import me.fullidle.ficore.ficore.common.api.pokemon.npc.PokeNPCEntityWrapperFactory;
import me.fullidle.ficore.ficore.common.api.pokemon.wrapper.*;
import me.fullidle.ficore.ficore.common.bukkit.entity.CraftEntity;
import net.minecraft.entity.LivingEntity;
import net.minecraft.entity.player.ServerPlayerEntity;
import org.bukkit.entity.Entity;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;
import java.util.stream.Collectors;

public class BattleManager implements IBattleManager<BattleController> {
    public static BattleManager INSTANCE = new BattleManager();

    /**
     * @throws RuntimeException 玩家已经在对局时或参与的宝可梦正在进化时，会报出
     */
    @NotNull
    @Override
    public IPokeBattle<BattleController> create(Player p1, Player p2) {
        val pp1 = new PlayerParticipant(((ServerPlayerEntity) CraftEntity.getHandle(p1)), createPlayerTeam(p1.getUniqueId()), 1);
        val pp2 = new PlayerParticipant(((ServerPlayerEntity) CraftEntity.getHandle(p2)), createPlayerTeam(p2.getUniqueId()), 1);
        val bc = createBattle(new BattleParticipant[]{pp1}, new BattleParticipant[]{pp2});
        return new PokeBattle(bc);
    }

    public static List<Pokemon> createPlayerTeam(UUID uuid) {
        val list = Lists.newArrayList(StorageProxy.getParty(uuid).getAll());
        list.removeIf(Objects::isNull);
        return list;
    }

    @Override
    public IPokeBattle<BattleController> create(Actor<?>[] side1, Actor<?>[] side2) {
        val s1 = new BattleParticipant[side1.length];
        val s2 = new BattleParticipant[side2.length];
        for (int i = 0; i < side1.length; i++)
            s1[i] = side1[i] == null ? null : ((BattleParticipant) side1[i].getOriginal());
        for (int i = 0; i < side2.length; i++)
            s2[i] = side2[i] == null ? null : ((BattleParticipant) side2[i].getOriginal());
        val bc = createBattle(s1, s2);
        return new PokeBattle(bc);
    }

    @Nullable
    @Override
    public IPokeBattle<BattleController> getBattle(Player player) {
        return getBattle((ServerPlayerEntity) CraftEntity.getHandle(player));
    }

    @Override
    public @NotNull IPokeBattle<BattleController> getBattle(Actor<?> actor) {
        return wrapper(((BattleParticipant) actor.getOriginal()).bc);
    }

    @Override
    public @NotNull IPokeBattle<BattleController> wrapper(BattleController battle) {
        return new PokeBattle(battle);
    }

    public IPokeBattle<BattleController> getBattle(ServerPlayerEntity player) {
        val battle = BattleRegistry.getBattle(player);
        return battle == null ? null : new PokeBattle(battle);
    }

    @Override
    public void createQuery(Player p1, Player p2) {
        val player1 = (ServerPlayerEntity) CraftEntity.getHandle(p1);
        val player2 = (ServerPlayerEntity) CraftEntity.getHandle(p2);
        String errorT = "玩家没有可参与的宝可梦!";
        val pokemon1 = Objects.requireNonNull(StorageProxy.getParty(player1).getFirstAblePokemon(), errorT);
        val pokemon2 = Objects.requireNonNull(StorageProxy.getParty(player2).getFirstAblePokemon(), errorT);
        new BattleQuery(
                player1,
                pokemon1.getOrSpawnPixelmon(player1),
                player2,
                pokemon2.getOrSpawnPixelmon(player2)
        );
    }

    @Override
    public ActorManager<?> getActorManager() {
        return V16ActorManager.INSTANCE;
    }

    public static class V16ActorManager implements ActorManager<BattleParticipant> {
        public static final V16ActorManager INSTANCE = new V16ActorManager();

        @Override
        public @NotNull Actor<?> create(Entity entity) throws UnsupportedOperationException {
            val version = V1_version.getInstance();
            val manager = (ActorManager<BattleParticipant>) version.getBattleManager().getActorManager();
            if (entity instanceof Player)
                return manager.wrap(new PlayerParticipant((ServerPlayerEntity) CraftEntity.getHandle(entity), createPlayerTeam(entity.getUniqueId()), 1));
            val pokeFactory = (PokeEntityWrapperFactory<PixelmonEntity>) version.getPokeEntityWrapperFactory();
            if (pokeFactory.isPokeEntity(entity))
                return manager.wrap(new WildPixelmonParticipant(pokeFactory.asPokeEntity(entity).getOriginal()));
            val npcFactory = (PokeNPCEntityWrapperFactory<NPCEntity>) version.getPokeNPCEntityWrapperFactory();
            if (npcFactory.isPokeNPCEntity(entity)) {
                val npc = npcFactory.asPokeNPCEntity(entity);
                if (npc.getOriginal() instanceof NPCTrainer)
                    return manager.wrap(new TrainerParticipant(((NPCTrainer) npc.getOriginal()), 1));
            }
            throw new UnsupportedOperationException("不支持的实体类型 >>" + entity.getType().name());
        }

        @Override
        public @NotNull Actor<BattleParticipant> wrap(BattleParticipant battleParticipant) {
            return new V16Actor(battleParticipant);
        }

        @Override
        public @Nullable Actor<?> getActor(Entity entity) {
            if (entity instanceof Player) {
                val player = (ServerPlayerEntity) CraftEntity.getHandle(entity);
                return wrap(BattleRegistry.getBattle(player).getPlayer(player));
            }
            val version = V1_version.getInstance();
            val pokeFactory = (PokeEntityWrapperFactory<PixelmonEntity>) version.getPokeEntityWrapperFactory();
            val npcFactory = (PokeNPCEntityWrapperFactory<NPCEntity>) version.getPokeNPCEntityWrapperFactory();
            if (pokeFactory.isPokeEntity(entity)) {
                val bc = pokeFactory.asPokeEntity(entity).getOriginal().battleController;
                return bc == null ? null : wrap(bc.getParticipantForEntity(((LivingEntity) CraftEntity.getHandle(entity))));
            }
            if (npcFactory.isPokeNPCEntity(entity)) {
                val npc = npcFactory.asPokeNPCEntity(entity).getOriginal();
                if (npc instanceof NPCTrainer) {
                    val bc = ((NPCTrainer) npc).battleController;
                    return bc == null ? null : wrap(bc.getParticipantForEntity(((LivingEntity) CraftEntity.getHandle(entity))));
                }
            }
            return null;
        }
    }

    public static class V16Actor extends Actor<BattleParticipant> {
        public V16Actor(BattleParticipant original) {
            super(original);
        }

        @Override
        public Entity getEntity() {
            val entity = getOriginal().getEntity();
            return entity == null ? null : CraftEntity.getEntity(entity);
        }

        @Override
        public Collection<IPokemonWrapper<?>> getTeam() {
            val list = new ArrayList<IPokemonWrapper<?>>();
            val factory = (IPokemonWrapperFactory<Pokemon>) V1_version.getInstance().getPokemonWrapperFactory();
            for (PixelmonWrapper wrapper : getOriginal().allPokemon) list.add(factory.create(wrapper.pokemon));
            return list;
        }

        @Override
        public Collection<IPokemonWrapper<?>> getCurrents() {
            val list = new ArrayList<IPokemonWrapper<?>>();
            val factory = (IPokemonWrapperFactory<Pokemon>) V1_version.getInstance().getPokemonWrapperFactory();
            for (PixelmonWrapper wrapper : getOriginal().controlledPokemon) list.add(factory.create(wrapper.pokemon));
            return list;
        }

        @Override
        public Class<BattleParticipant> getType() {
            return BattleParticipant.class;
        }
    }

    @Getter
    public static class PokeBattle extends IPokeBattle<BattleController> {
        public PokeBattle(BattleController bc) {
            super(bc);
        }

        @Override
        public Collection<Player> getPlayers() {
            return this.getOriginal().getPlayers()
                    .stream()
                    .map(s -> ((Player) CraftEntity.getEntity(s.player)))
                    .collect(Collectors.toList());
        }

        @Override
        public Collection<Actor<?>> getPlayerActors() {
            val actorManager = (ActorManager<BattleParticipant>)FIData.V1_version.getBattleManager().getActorManager();
            return getOriginal().getPlayers().stream().map(actorManager::wrap).collect(Collectors.toList());
        }

        @Override
        public Collection<Actor<?>> getActors() {
            return Collections.emptyList();
        }

        @Override
        public void end() {
            this.getOriginal().endBattle();
        }

        @Override
        public void start() {
            startBattle(this.getOriginal());
        }

        @Override
        public Class<BattleController> getType() {
            return BattleController.class;
        }

    }

    /**
     * original
     *
     * @see BattleRegistry#startBattle(BattleParticipant[], BattleParticipant[], BattleRules)
     */
    @SneakyThrows
    @NotNull
    public static BattleController createBattle(BattleParticipant[] team1, BattleParticipant[] team2) {
        BattleController battle = new BattleController(team1, team2, new BattleRules(BattleType.SINGLE));
        val method = BattleRegistry.class.getDeclaredMethod("canParticipantsBattle", BattleController.class);
        method.setAccessible(true);
        if (!(boolean) method.invoke(null, battle)) {
            throw new RuntimeException("Players cannot battle");
        }
        return battle;
    }

    @SneakyThrows
    public static void startBattle(BattleController battle) {
        val pp = battle.participants.get(0);
        val method = BattleRegistry.class.getDeclaredMethod("performPreBattleChecksAndEvents", BattleController.class, BattleParticipant[].class, BattleParticipant[].class);
        method.setAccessible(true);
        val array1 = battle.getTeam(pp).toArray(new BattleParticipant[0]);
        val array2 = battle.getOpponents(pp).toArray(new BattleParticipant[0]);
        val invoke = (boolean) method.invoke(null, battle, array1, array2);
        if (!invoke) return;
        BattleRegistry.registerBattle(battle);
        Pixelmon.EVENT_BUS.post(new BattleStartedEvent.Post(battle, array1, array2));
    }
}

File: v1_21/Cob/src/main/kotlin/me/fullidle/ficore/ficore/vCob_21/BattleManager.kt
package me.fullidle.ficore.ficore.vCob_21

import com.cobblemon.mod.common.api.battles.model.PokemonBattle
import com.cobblemon.mod.common.api.battles.model.actor.BattleActor
import com.cobblemon.mod.common.api.battles.model.actor.EntityBackedBattleActor
import com.cobblemon.mod.common.api.events.CobblemonEvents
import com.cobblemon.mod.common.api.events.battles.BattleStartedEvent
import com.cobblemon.mod.common.battles.BattleFormat
import com.cobblemon.mod.common.battles.BattleRegistry
import com.cobblemon.mod.common.battles.BattleSide
import com.cobblemon.mod.common.battles.actor.PlayerBattleActor
import com.cobblemon.mod.common.battles.actor.PokemonBattleActor
import com.cobblemon.mod.common.battles.pokemon.BattlePokemon
import com.cobblemon.mod.common.entity.npc.NPCBattleActor
import com.cobblemon.mod.common.entity.npc.NPCEntity
import com.cobblemon.mod.common.entity.pokemon.PokemonEntity
import com.cobblemon.mod.common.pokemon.Pokemon
import com.cobblemon.mod.common.util.getPlayer
import com.cobblemon.mod.common.util.party
import me.fullidle.ficore.ficore.common.V1_version
import me.fullidle.ficore.ficore.common.api.pokemon.battle.IBattleManager
import me.fullidle.ficore.ficore.common.api.pokemon.battle.IPokeBattle
import me.fullidle.ficore.ficore.common.api.pokemon.battle.actor.Actor
import me.fullidle.ficore.ficore.common.api.pokemon.battle.actor.ActorManager
import me.fullidle.ficore.ficore.common.api.pokemon.wrapper.IPokemonWrapper
import me.fullidle.ficore.ficore.common.api.pokemon.wrapper.IPokemonWrapperFactory
import me.fullidle.ficore.ficore.common.api.pokemon.wrapper.PokeEntityWrapperFactory
import me.fullidle.ficore.ficore.common.api.pokemon.wrapper.PokeNPCEntityWrapperFactory
import me.fullidle.ficore.ficore.common.bukkit.entity.CraftEntity
import net.minecraft.world.entity.LivingEntity
import org.bukkit.Bukkit
import org.bukkit.entity.Entity
import org.bukkit.entity.Player
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import kotlin.reflect.KProperty1
import kotlin.reflect.full.declaredMemberFunctions
import kotlin.reflect.full.declaredMemberProperties
import kotlin.reflect.jvm.isAccessible

object BattleManager : IBattleManager<PokemonBattle> {
    override fun create(
        p1: Player,
        p2: Player
    ): IPokeBattle<PokemonBattle> {
        val uuid1 = p1.uniqueId
        val uuid2 = p2.uniqueId
        val team1 = uuid1.getPlayer()!!.party().toBattleTeam()
        val team2 = uuid2.getPlayer()!!.party().toBattleTeam()
        val pb = PokemonBattle(
            BattleFormat.GEN_9_SINGLES,
            BattleSide(PlayerBattleActor(uuid1, team1)),
            BattleSide(PlayerBattleActor(uuid2, team2))
        )
        return PokeBattle(pb)
    }

    override fun create(
        side1: Array<out Actor<*>>,
        side2: Array<out Actor<*>>
    ): IPokeBattle<PokemonBattle> {
        val pb = PokemonBattle(
            BattleFormat.GEN_9_SINGLES,
            BattleSide(*side1.map { it.original as BattleActor }.toTypedArray()),
            BattleSide(*side2.map { it.original as BattleActor }.toTypedArray())
        )
        return PokeBattle(pb)
    }

    override fun getBattle(player: Player): IPokeBattle<PokemonBattle>? {
        return BattleRegistry.getBattleByParticipatingPlayerId(player.uniqueId)?.let {
            PokeBattle(it)
        }
    }

    override fun getBattle(actor: Actor<*>): IPokeBattle<PokemonBattle> {
        return wrapper((actor.original as BattleActor).battle)
    }

    override fun wrapper(battle: PokemonBattle): IPokeBattle<PokemonBattle> {
        return PokeBattle(battle)
    }

    override fun getActorManager(): ActorManager<*> {
        return CobActorManager
    }

    object CobActorManager : ActorManager<BattleActor> {
        override fun create(entity: Entity): Actor<*> {
            val pokemonFactory =
                V1_version.getInstance().pokemonWrapperFactory as PokeEntityWrapperFactory<PokemonEntity>
            val npcFactory =
                V1_version.getInstance().pokeNPCEntityWrapperFactory as PokeNPCEntityWrapperFactory<NPCEntity>

            return CobActor(
                when {
                    entity is Player -> {
                        entity.uniqueId.let {
                            PlayerBattleActor(it, it.getPlayer()!!.party().toBattleTeam())
                        }
                    }

                    pokemonFactory.isPokeEntity(entity) -> {
                        PokemonBattleActor(
                            entity.uniqueId,
                            BattlePokemon.safeCopyOf((pokemonFactory.asPokeEntity(entity).original as PokemonEntity).pokemon),
                            16.0f
                        )
                    }

                    npcFactory.isPokeNPCEntity(entity) -> {
                        (npcFactory.asPokeNPCEntity(entity).original as NPCEntity).let {
                            NPCBattleActor(it, it.party ?: throw UnsupportedOperationException("不支持的npc 无队伍"), it.skill ?: 1)
                        }
                    }

                    else -> throw UnsupportedOperationException("不支持的实体类型 >> ${entity.type.name}")
                }
            )
        }

        override fun wrap(t: BattleActor): Actor<BattleActor> {
            return CobActor(t)
        }

        override fun getActor(entity: Entity): Actor<*>? {
            val pokemonFactory =
                V1_version.getInstance().pokemonWrapperFactory as PokeEntityWrapperFactory<PokemonEntity>
            val npcFactory =
                V1_version.getInstance().pokeNPCEntityWrapperFactory as PokeNPCEntityWrapperFactory<NPCEntity>
            return when {
                entity is Player -> {
                    entity.uniqueId.let {
                        BattleRegistry.getBattleByParticipatingPlayerId(it)?.getActor(it)
                    }
                }

                pokemonFactory.isPokeEntity(entity) -> {
                    (pokemonFactory.asPokeEntity(entity).original as PokemonEntity).let { pe ->
                        pe.battle?.getActor(pe.uuid)
                    }
                }

                npcFactory.isPokeNPCEntity(entity) -> {
                    (npcFactory.asPokeNPCEntity(entity).original as NPCEntity).let { npc->
                        npc.battleIds.firstOrNull()?.let {
                            BattleRegistry.getBattle(it)?.getActor(npc.uuid)
                        }
                    }
                }

                else -> null
            }?.let {
                wrap(it)
            }
        }
    }

    class CobActor(original: BattleActor) : Actor<BattleActor>(original) {
        override fun getEntity(): Entity? {
            return (original as? EntityBackedBattleActor<LivingEntity>)?.entity?.let {
                CraftEntity.getEntity(it)
            }
        }

        override fun getTeam(): Collection<IPokemonWrapper<*>> {
            return (V1_version.getInstance().pokemonWrapperFactory as IPokemonWrapperFactory<Pokemon>).let { factory ->
                original.pokemonList.map {
                    factory.create(it.originalPokemon)
                }
            }
        }

        override fun getCurrents(): Collection<IPokemonWrapper<*>> {
            return (V1_version.getInstance().pokemonWrapperFactory as IPokemonWrapperFactory<Pokemon>).let { factory ->
                original.activePokemon.mapNotNull { ap ->
                    ap.battlePokemon?.let {
                        factory.create(it.originalPokemon);
                    }
                }
            }
        }

        override fun getType(): Class<BattleActor> {
            return BattleActor::class.java
        }
    }

    class PokeBattle(
        original: PokemonBattle
    ) : IPokeBattle<PokemonBattle>(original) {
        override fun getPlayers(): Collection<Player>? {
            return original.sides.flatMap { side ->
                side.actors
                    .filterIsInstance<PlayerBattleActor>()
                    .map { actor -> Bukkit.getPlayer(actor.uuid)!! }
            }
        }

        override fun getPlayerActors(): Collection<Actor<*>> {
            return (V1_version.getInstance().battleManager.actorManager as ActorManager<BattleActor>).let { manager ->
                original.players.mapNotNull { player ->
                    original.getActor(player)?.let {
                        manager.wrap(it)
                    }
                }
            }
        }

        override fun getActors(): Collection<Actor<*>> {
            return (V1_version.getInstance().battleManager.actorManager as ActorManager<BattleActor>).let { manager ->
                original.actors.map {
                    manager.wrap(it)
                }
            }
        }

        override fun end() {
            original.end()
        }

        override fun start() {
            val start: () -> Unit = {
                (BattleRegistry::class.declaredMemberProperties.find {
                    it.name == "battleMap" && it.returnType == ConcurrentHashMap::class
                }!! as KProperty1<BattleRegistry, ConcurrentHashMap<UUID, PokemonBattle>>).apply {
                    this.isAccessible = true
                }.get(BattleRegistry)[original.battleId] = original
                BattleRegistry::class.declaredMemberFunctions.find {
                    it.name == "startShowdown"
                }!!.call(original)
            }

            val preBattleEvent = BattleStartedEvent.Pre(original)
            CobblemonEvents.BATTLE_STARTED_PRE.postThen(preBattleEvent) {
                start()
                CobblemonEvents.BATTLE_STARTED_POST.post(BattleStartedEvent.Post(original))
                return
            }
            return
        }

        override fun getOriginal(): PokemonBattle {
            return original
        }

        override fun getType(): Class<PokemonBattle> {
            return PokemonBattle::class.java
        }
    }
}
